### PR TITLE
Add a configurable randomness source

### PR DESCRIPTION
**Pull request by rhernandez35**
I had two goals for this change:

1. Expose the configuration option without exposing any new classes or
interfaces. We should be in strict control of possible randomness
sources so we can guarantee that they actually do provide random values
and that they provide randomness efficiently. I didn't want it to be
possible for users to tank performance by inadvertently using a single
`Random` instance.
2. Make the random source configurable and scoped to each
`AWSXRayRecorder` instance. I didn't want to introduce any more global
shared state, so this unfortunately meant I wasn't able to remove the
global SecureRandom usage in two places: `TraceId#fromString` and
`DefaultThrowableSerializationStrategy#describeInContext`. The latter is
at least aware of the local segment context and will try to use the
random generator from its creator when possible, but the former is
unresolvable without either adding more global shared state or exposing
the `RandomGenerator` class.

This commit adds the following public APIs:

* `AWSXRayRecorder#useFastRandomGenerator` and
`AWSXRayRecorder#useSecureRandomGenerator`: for configuring the behavior
of each `AWSXRayRecorder` instance.
* `AWSXRayRecorderBuilder#useFastRandomGenerator` and
`AWSXRayRecorderBuilder#useSecureRandomGenerator`: same as above
* `ThrowableSerializationStrategy#describeInContext(SegmentContext,
Throwable, List<Subsegment>)`: a new method that allows the current
segment context to be provided to `ThrowableSerializationStrategy`. This
is currently only useful for getting the ID generator from the current
segment's creator, but I can imagine there being other uses for this
later.

I unfortunately had to add two internal-only public methods to
`AWSXRayRecorder` for actually getting random values from its configured
random generator. Adding methods there allowed me to keep the
`RandomGenerator` class hidden from users.

There are some other minor changes in this PR:

* Removed quadratic String concatenation in `Entity#generateId`
* `SegmentContextResolverChain` no longer double-resolves contexts

*Issue #, if available:*
#216 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/ThrowableSerializationStrategy.java
Comment (willarmiros) [inline aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/ThrowableSerializationStrategy.java:51]: Just to play devil's advocate since this is a fairly tedious change and new API, do we need the performance bump in the throwable serializer? It's only used for requests that are already "bad" by nature of having an exception that needs recording. 

I'm ok with leaving it, especially if you think the performance is still important here, but will leave it up to you.

Comment (willarmiros) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/ThrowableSerializationStrategy.java:51]: Actually, looking at it again, is the segment context (and this new API) necessary at all? Can't you just grab a subsegment out of the list in the original API, do `getCreator()` on it and use the `generateId` method there?

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/AWSXRayRecorder.java
Comment (willarmiros) [inline aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/AWSXRayRecorder.java:1062]: I know @rhernandez35 you want to avoid exposing these, but I'm not sure how I feel about adding new internal-only methods to an already very bloated Recorder class. IMO these methods can be extracted into a public interface/class (like `RandomGenerator`) and exposed to customers. That is to some degree what we're requesting OpenTelemetry, a pluggable ID generation experience. I'm not sure if customers would bother using it unless they really knew what they were doing, and we can still guarantee randomness with our *own* generators.

@anuraaga I'd appreciate you weighing in on this too, I could be completely off the mark here.

**Review by willarmiros — COMMENTED** [review]
Thanks again for the changes! Left some thoughts and we should prob wait to hear from @anuraaga before getting this in.

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/AWSXRayRecorder.java
Comment (rhernandez35) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/AWSXRayRecorder.java:1062]: If you're fine with increasing the surface area then I'll gladly do it since the code would be so much cleaner. However, my goal was to avoid providing customers a way to evade our curated selection of randomness providers. I really don't want someone to think "oh hey I can save memory by just using `Math#random`" and accidentally add a global synchronization point on every new ID creation.

Adding these methods here was definitely a gigantic hack, but it seemed like the lesser of two evils.

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/ThrowableSerializationStrategy.java
Comment (rhernandez35) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/ThrowableSerializationStrategy.java:51]: I did that at first, but there's no guarantee that the list will contain any subsegments. In my simple demo service, we just have the one master segment. I could search the list as a fallback but I didn't think that was necessary since I'm hoping most users are using the default serializer, which is able to always provide the segment in use.

I also experimented with adding an `Entity` parameter. Maybe I should have kept that? It was definitely safer – the only way to reach this method in X-Ray's codebase is through `EntityImpl#addException`, and while extremely unlikely, there's no guarantee that the entity invoking `describeInContext` is the same entity that's stored by the segment context. This wasn't a problem with the `Entity` parameter because `EntityImpl` could just give `this` as an argument.

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/AWSXRayRecorderBuilder.java
Comment (rhernandez35) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/AWSXRayRecorderBuilder.java:232]: indeed we can

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/AWSXRayRecorder.java
Comment (anuraaga) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/AWSXRayRecorder.java:1062]: I lean towards having only two types of random - OpenTelemetry supports trace ID customization since it needs to support all backends, including weird ones like X-Ray. X-Ray SDK on the flip side can't be used with any other backend so we can just limit the options.

Instead of the pattern of deprecated method for internal API, we have the `.internal` package - can we move he internal pieces there?

**Review by anuraaga — COMMENTED** [review]
Sorry for the late review - left a high level comment about the internal API. Thanks a lot for the help @rhernandez35!

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/AWSXRayRecorder.java
Comment (rhernandez35) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/AWSXRayRecorder.java:1062]: sure, I'm fine with that if you are.

I can say from experience though that users don't really care what the fully qualified name for a class is when they use it. If it's pubic and their IDE can autoresolve it, they often don't even know what package it's in. If it's public, it _will_ be used in ways you can't anticipate.

Comment (willarmiros) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/AWSXRayRecorder.java:1062]: I think the `internal` package sounds good to me. Sure customers *can* override it if they want, but it will not be a publicly documented configuration like the SDK's other configs (as a matter of fact the JavaDoc should specifically call out that it's an internal-only API that is not meant to be used by consumers of the SDK, which may help guide some of those IDE users 😉 ).

**Comment by rhernandez35** [comment]
OK, I moved the random generators to `internal` and added a getter in `AWSXRayRecorder`. Everything else is mostly the same as in the last revision.

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/utils/SecureRandomSupplier.java
Comment (anuraaga) [inline aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/utils/SecureRandomSupplier.java:31]: We removed threadlocal securerandom here after adding benchmarks for it in #162

I believe it's because the OS's random is guaranteed to be a global lock, so a thread-local securerandom instance doesn't reduce locking but does add overhead for the threadlocal lookup.

If this makes sense (maybe my benchmark is missing something, never know), could you revert this change?

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/internal/RandomGenerator.java
Comment (anuraaga) [inline aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/internal/RandomGenerator.java:35]: Since this is an internal class we can tie it to our requirements - let's have it return the base16 string itself. For secure random, `nextBytes` is much faster but for normal random, two `nextLong`s is faster, we should take advantage of that. If it's too much for this PR, we can address it later though no worries.

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/AWSXRayRecorder.java
Comment (rhernandez35) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/AWSXRayRecorder.java:952]: Ok, I'll tone it down a bit.

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/utils/SecureRandomSupplier.java
Comment (rhernandez35) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/utils/SecureRandomSupplier.java:31]: the entropy pool is just one source of blocking and actually isn't directly queried all that often (usually only for seed generation). the main source of contention I'm worried about is the thread-safety of the RNG algorithm itself – RNGs typically maintain internal state that prevent concurrent reads/updates. This blob of code checks for when that's true and uses thread locals as necessary.

what os/arch did you run your benchmarks on? what crypto provider(s) did you install? which JVM – corretto? [AmazonCorrettoCryptoProvider](https://aws.amazon.com/about-aws/whats-new/2019/07/introducing-the-amazon-corretto-crypto-provider/) provides a thread-safe `SecureRandom` implementation that stores its state in thread-locals, so `SecureRandom` skips synchronizing on calls into it. You'll see lock contention in SecureRandom if you use a crypto provider that doesn't provide a thread-safe random.

Comment (rhernandez35) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/utils/SecureRandomSupplier.java:31]: now that I think about it, it would probably be better to always use a thread-local SecureRandom since "thread-safe" could also be implemented by just slapping `synchronized` on your random spi's methods.

Comment (anuraaga) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/utils/SecureRandomSupplier.java:31]: It sounds like coretto would be even faster without the threadlocal. But I just tested with normal openjdk 14 on Mac and even with that, not using threadlocal was a bit faster. I don't know how well my theory checks out, but since the benchmark numbers aren't convincing and threadlocal securerandom adds complexity, let's go ahead and stick with the singleton please. Thanks!

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/ThrowableSerializationStrategy.java
Comment (willarmiros) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/ThrowableSerializationStrategy.java:51]: Ok yeah I like passing the entity instead of the context in that case - if we're just using the context to retrieve the entity then it definitely sounds simpler and safer to pass in the entity.

**Review by willarmiros — COMMENTED** [review]
Couple more comments

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/ThreadLocalStorage.java
Comment (rhernandez35) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/ThreadLocalStorage.java:28]: I suppose, but this is documented as an internal-only class. I thought y'all were ok with making breaking changes to internal code?

Comment (anuraaga) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/ThreadLocalStorage.java:28]: Ah good catch - yeah let's not change to final this time. Internal was added relatively recently in the hopes of significant cleanup in a major version like hiding this class. So the internal package is ok, all classes there started out as that, but not anything outside. If this code generally looks like a mess, yes you're right :)

**Review by willarmiros — APPROVED** [review]
LGTM, will wait on @anuraaga to approve and merge. Thanks for your contribution and patience @rhernandez35!

File: aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/DefaultThrowableSerializationStrategy.java
Comment (rhernandez35) [inline aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/DefaultThrowableSerializationStrategy.java:136]: how would you feel about me using the ID generator from the global recorder (via `AWSXRay::getGlobalRecorder`) instead of this?

Comment (rhernandez35) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/DefaultThrowableSerializationStrategy.java:136]: and same for in `TraceID::fromString`/`TraceID::create`

Comment (willarmiros) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/DefaultThrowableSerializationStrategy.java:136]: Ah I'd actually prefer that! Since that will also default to secure if the user hasn't configured their recorder, but at least has a chance of picking up the configured value if the first two conditions don't get hit somehow.

Comment (rhernandez35) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/DefaultThrowableSerializationStrategy.java:136]: Are users allowed to set the global recorder to `null`? Most code assumes it won't return null, but there's no explicit check in `setGlobalRecorder`. Further, there's a test that even sets it to null, alongside some code in `TracingHandler` that's meant to deal with `getGlobalRecorder` returning null (although this code also assumes that all subsequent invocations of `getGlobalRecorder` will return an object).

Comment (willarmiros) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/DefaultThrowableSerializationStrategy.java]: We added `Nullable` annotations throughout the SDK and left the `globalRecorder` field and corresponding getter in `AWSXRay` as not `Nullable` so I think it's safe to assume that customers are respecting the non-null contract. In `TracingHandler` I think we're protecting against the `TracingHandler` being constructed with a `null` recorder parameter, and then assume that `getGlobalRecorder` will provide the recorder if so. What test sets it to null?

Comment (rhernandez35) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/DefaultThrowableSerializationStrategy.java]: `testRaceConditionOnRecorderInitialization` in `TracingHandlerTest`

Comment (willarmiros) [reply aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/DefaultThrowableSerializationStrategy.java]: Ah yeah that's unrelated to testing if the customer has set the global recorder to null, it's [this issue](https://github.com/aws/aws-xray-sdk-java/issues/29) where the AWS SDK's auto-instrumentor would create a `TracingHandler` and call `AWSXRay.getGlobalRecorder` in parallel with the recorder's start up before it had finished initializing. Not something that we'd expect in regular customer code that's run after the recorder is initialized.

**Review by anuraaga — APPROVED** [review]
Thanks @rhernandez35!

**Pull request merged**
